### PR TITLE
feat: Alloy migration

### PR DESCRIPTION
**Pull request by Evalir**
Migrates the repo to alloy. The only remaining ethers types is `Chain`, which is still being worked on as a separate package.

### CODE REVIEW

File: src/block_number.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
+use alloy_primitives::U64;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::{fmt, str::FromStr};
+
+/// A block number or tag.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, Hash)]
```
Comment (Evalir) [inline src/block_number.rs:6]: This eventually should come from RPC types.

File: src/serde_helpers.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
+use crate::block_number::BlockNumber;
+use alloy_primitives::{U256, U64};
+use serde::{Deserialize, Deserializer};
+use std::{
+    convert::{TryFrom, TryInto},
+    str::FromStr,
+};
+
+/// Helper type to parse numeric strings, `u64` and `U256`
+#[derive(Deserialize, Debug, Clone)]
```
Comment (Evalir) [inline src/serde_helpers.rs:10]: These were moved from `ethers_core`

File: src/units.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
+use crate::utils::ConversionError;
+use std::{convert::TryFrom, fmt, str::FromStr};
+
+/// Common Ethereum unit types.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Units {
```
Comment (Evalir) [inline src/units.rs:6]: Also moved from `ethers_core`. This feels like it should live somewhere else (maybe alloy), as a primitive.

**Review by mattsse — COMMENTED** [review]
this adds a bunch of rpc junk that should live somewhere else, alloy-rpc-types I think.

so perhaps we do that first?

but also fine with doing some duplication here in the meantime

File: src/block_number.rs (hunk 1 above)
Comment (gakonst) [reply src/block_number.rs:6]: Flagging this also exists here: https://github.com/paradigmxyz/reth/blob/529635f8d4d627d3652807e3f0377b3c256de4fc/crates/primitives/src/block.rs#L548.

File: src/contract.rs [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
     utils::{deserialize_address_opt, deserialize_source_code},
     Client, EtherscanError, Response, Result,
 };
-use ethers_core::{
-    abi::{Abi, Address, RawAbi},
-    types::{serde_helpers::deserialize_stringified_u64, Bytes},
-};
+use alloy_json_abi::JsonAbi as Abi;
+use alloy_primitives::{Address, Bytes};
+use ethers_core::abi::RawAbi;
```
Comment (gakonst) [inline src/contract.rs:9]: This should be removed and adjusted to the right Alloy type, am OK w/ doing in follow up.

File: src/serde_helpers.rs (hunk 2 above)
Comment (gakonst) [reply src/serde_helpers.rs:10]: How do we do this in Reth? @mattsse

File: src/units.rs (hunk 3 above)
Comment (gakonst) [reply src/units.rs:6]: @prestwich @DaniPopes I am OK w/ moving this to `core/` but no strong opinions

File: tests/it/contract.rs [hunk 5]
```diff
@@ -99,7 +102,10 @@ async fn can_fetch_contract_source_tree_for_singleton_contract() {
             SourceCodeMetadata::SourceCode(_)
         ));
         assert_eq!(item.source_code.sources().len(), 1);
-        assert_eq!(item.abi().unwrap(), serde_json::from_str(DAO_ABI).unwrap());
+        assert_eq!(
+            item.abi().unwrap(),
+            serde_json::from_str(DEPOSIT_CONTRACT_ABI).unwrap()
```
Comment (gakonst) [inline tests/it/contract.rs:107]: did anything not work with the DAO ABI?

**Review by gakonst — APPROVED** [review]
Good w/ me. The follow-up should remove RawAbi.

**Pull request merged**